crucible-symio/src/Lang/Crucible/SymIO/Types.hs

```diff
@@ -0,0 +1,208 @@
+
+pattern FileSystemRepr :: () => (1 <= w, ty ~ FileSystemType w) => NatRepr w -> TypeRepr ty
+pattern FileSystemRepr w <- IntrinsicRepr (testEquality (knownSymbol :: SymbolRepr "VFS_filesystem") -> Just Refl)
+                                           (Empty :> BVRepr w)
+  where
+    FileSystemRepr w = IntrinsicRepr knownSymbol (Empty :> BVRepr w)
+
+-- | The crucible type of file handles.
+type FileHandleType w = ReferenceType (MaybeType (FilePointerType w))
+
+-- |  A file handle is a mutable file pointer that increments every time it is read.
+type FileHandle sym w = RegValue sym (FileHandleType w)
+
+-- | A 'File' represents a file in the filesystem independent
+-- of any open handles to it
+data File sym w = File (NatRepr w) (SymInteger sym)
+
+pattern FileRepr :: () => (1 <= w, ty ~ FileType w) => NatRepr w -> TypeRepr ty
+pattern FileRepr w <- IntrinsicRepr (testEquality (knownSymbol :: SymbolRepr "VFS_file") -> Just Refl)
+                                           (Empty :> BVRepr w)
+  where
+    FileRepr w = IntrinsicRepr knownSymbol (Empty :> BVRepr w)
+
+-- | The crucible-level type of 'File'
+type FileType w = IntrinsicType "VFS_file" (EmptyCtx ::> BVType w)
+
+instance (IsSymInterface sym) => IntrinsicClass sym "VFS_file" where
+  type Intrinsic sym "VFS_file" (EmptyCtx ::> BVType w) = File sym w
+
+  muxIntrinsic sym _iTypes _nm (Empty :> BVRepr _w) = muxFile sym
+  muxIntrinsic _ _ nm ctx = typeError nm ctx
+
+muxFile ::
+  IsSymInterface sym =>
+  sym ->
+  Pred sym ->
+  File sym w ->
+  File sym w ->
+  IO (File sym w)
+muxFile sym p (File w f1) (File _w f2) = File w <$> baseTypeIte sym p f1 f2
+
+-- | A file pointer represents an index into a particular file.
+data FilePointer sym w =
+  FilePointer (File sym w) (SymBV sym w) 
+
+-- | The crucible type of 'FilePointer'
+type FilePointerType w = IntrinsicType "VFS_filepointer" (EmptyCtx ::> BVType w)
+
+instance (IsSymInterface sym) => IntrinsicClass sym "VFS_filepointer" where
+  type Intrinsic sym "VFS_filepointer" (EmptyCtx ::> BVType w) = FilePointer sym w
+
+  muxIntrinsic sym _iTypes _nm (Empty :> (BVRepr _w)) = muxFilePointer sym
+  muxIntrinsic _ _ nm ctx = typeError nm ctx
+
+-- | Mux on 'FilePointer'
+muxFilePointer ::
+  (1 <= w) =>
+  IsSymInterface sym =>
+  sym ->
+  Pred sym ->
+  FilePointer sym w ->
+  FilePointer sym w ->
+  IO (FilePointer sym w)
+muxFilePointer sym p (FilePointer f1 off1) (FilePointer f2 off2) =
+  do b   <- muxFile sym p f1 f2
+     off <- bvIte sym p off1 off2
+     return $ FilePointer b off
+
+
+type DataChunk sym w = CA.ArrayChunk sym (BaseBVType w) (BaseBVType 8)
+
+type SizedDataChunkType w = SymbolicStructType (EmptyCtx ::> BaseArrayType (EmptyCtx ::> BaseBVType w) (BaseBVType 8) ::> BaseBVType w)
+type SizedDataChunk sym w = SymStruct sym (EmptyCtx ::> BaseArrayType (EmptyCtx ::> BaseBVType w) (BaseBVType 8) ::> BaseBVType w)
+
+
+pattern FileHandleRepr :: () => (1 <= w, ty ~ FileHandleType w) => NatRepr w -> TypeRepr ty
```

Comment:
I think this is worth a comment.  Especially, it's not immediately clear what is meant if the reference contains `Nothing`. I gather this is the state of a closed handle, but I'm not 100% sure.

Reply:
This is correct, a `FileHandle` that points to `Nothing` simply represents a closed handle.

Reply:
Added a bit of detail (and some additional comments on some of the related types)